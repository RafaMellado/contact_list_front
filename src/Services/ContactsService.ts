import {
  Contact,
  ContactRequestBody,
  ContactWithContactHistories,
} from "./Interfaces/Contact";
import ContactsRepository from "./Repository/ContactsRepository";

class ContactsService {
  index(params: object): Promise<Contact[]> {
    return ContactsRepository.index(params);
  }

  show(id: number): Promise<ContactWithContactHistories> {
    return ContactsRepository.show(id);
  }

  create(data: ContactRequestBody): Promise<Contact> {
    return ContactsRepository.create(data);
  }

  update(id: number, data: ContactRequestBody): Promise<Contact> {
    return ContactsRepository.update(id, data);
  }

  delete(id: number): Promise<Response> {
    return ContactsRepository.delete(id);
  }
}

export default new ContactsService();
