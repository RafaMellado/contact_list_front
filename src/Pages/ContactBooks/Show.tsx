import { useEffect, useState } from "react";
import { Button, Col, Row } from "react-bootstrap";
import { useTranslation } from "react-i18next";
import { useNavigate, useParams } from "react-router-dom";
import ContactBooksService from "../../Services/ContactBooksService";
import { ContactBookWithContacts } from "../../Services/Interfaces/ContactBook";
import { ContactBookContactCard } from "../../Components/ContactBookContactCard";
import ContactsService from "../../Services/ContactsService";

export function ContactBookShow() {
  const [contactBook, setContactBook] = useState<ContactBookWithContacts>();

  const { id } = useParams();
  const { t } = useTranslation();

  const navigate = useNavigate();

  const TRANSLATIONS: string = "contactBooksShow";

  useEffect(() => {
    ContactBooksService.show(Number(id)).then((response) =>
      setContactBook(response)
    );
  }, [id]);

  const addContact = () => {
    navigate(`/contact-book/${id}/contact/new`);
  };

  const backBtn = () => {
    navigate("/home");
  };

  const goToContactShow = (contactId: number) => {
    navigate(`/contact/${contactId}`);
  };

  const goToEditContact = (contactId: number) => {
    navigate(`/contact/${contactId}/edit`);
  };

  const deleteContact = async (id: number) => {
    await ContactsService.delete(id);

    reassignContacts(id);
  };

  const reassignContacts = (deletedId: number) => {
    const updatedContactBook = {
      ...contactBook,
      contacts: contactBook?.contacts.filter((item) => item.id !== deletedId),
    };

    setContactBook(updatedContactBook as ContactBookWithContacts);
  };

  return (
    <Row>
      <Col>
        <h2>{contactBook?.name}</h2>

        <Button className="me-2" onClick={backBtn}>
          {t<string>(`${TRANSLATIONS}.backBtn`)}
        </Button>

        <Button onClick={addContact}>
          {t<string>(`${TRANSLATIONS}.addContactBtn`)}
        </Button>

        <h3>{t<string>(`${TRANSLATIONS}.contacts`)}</h3>

        <Row>
          {contactBook?.contacts.map((item) => {
            return (
              <Col xs={12} sm={6} md={3} key={item.id}>
                <ContactBookContactCard
                  item={item}
                  cardFn={() => goToContactShow(item.id)}
                  editFn={() => goToEditContact(item.id)}
                  deleteFn={() => deleteContact(item.id)}
                />
              </Col>
            );
          })}
        </Row>
      </Col>
    </Row>
  );
}
