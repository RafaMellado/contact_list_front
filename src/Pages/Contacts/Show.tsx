import { useEffect, useState } from "react";
import { Button, Col, Row } from "react-bootstrap";
import { useTranslation } from "react-i18next";
import { useNavigate, useParams } from "react-router-dom";
import ContactsService from "../../Services/ContactsService";
import { ContactWithContactHistories } from "../../Services/Interfaces/Contact";
import { ContactCard } from "../../Components/ContactCard";
import { ContactHistoryCard } from "../../Components/ContactHistoryCard";

export function ContactShow() {
  const [contact, setContact] = useState<ContactWithContactHistories>();

  const { id } = useParams();
  const { t } = useTranslation();

  const navigate = useNavigate();

  const TRANSLATIONS: string = "contactsShow";

  useEffect(() => {
    ContactsService.show(Number(id)).then((response) => setContact(response));
  }, [id]);

  const backBtn = (id: number) => {
    navigate(`/contact-book/show/${id}`);
  };

  const editContact = () => {
    navigate(`/contact/${id}/edit`);
  };

  return (
    <Row>
      <Col xs={12} md={6}>
        {contact && (
          <ContactCard item={contact}>
            <>
              <Button
                className="me-2"
                onClick={() => backBtn(Number(contact.contact_book_id))}
              >
                {t<string>(`${TRANSLATIONS}.backBtn`)}
              </Button>

              <Button className="me-2" onClick={editContact}>
                {t<string>(`${TRANSLATIONS}.editContact`)}
              </Button>
            </>
          </ContactCard>
        )}
      </Col>

      <Col xs={12} md={6}>
        {!!contact?.contact_histories.length && (
          <>
            <h2>{t<string>(`${TRANSLATIONS}.history`)}</h2>

            {contact.contact_histories.map((item, index) => {
              return (
                <ContactHistoryCard
                  item={item}
                  key={index}
                  translations={TRANSLATIONS}
                />
              );
            })}
          </>
        )}
      </Col>
    </Row>
  );
}
