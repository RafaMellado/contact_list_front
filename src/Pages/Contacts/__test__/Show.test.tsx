import { ContactShow } from "../Show";
import { act, fireEvent, render, waitFor } from "@testing-library/react";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import ContactsService from "../../../Services/ContactsService";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({ t: (key: string) => key }),
}));

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({ id: 1 }),
}));

jest.mock("../../../Services/ContactsService");

const mockedContactService = ContactsService as jest.Mocked<
  typeof ContactsService
>;

describe(`<${ContactShow.name} />`, () => {
  const history = createMemoryHistory();

  const factoryComponent = async () => {
    const component = render(
      <Router location="/" navigator={history}>
        <ContactShow />
      </Router>
    );

    await waitFor(() => {
      expect(component.getByTestId("contact-history-card")).toBeInTheDocument();
    });

    return component;
  };

  beforeEach(() => {
    mockedContactService.show.mockResolvedValue({
      id: 1,
      givenname: "test",
      surname: "surname test",
      email: "test@email.es",
      phone: "12345",
      contact_book_id: 1,
      contact_histories: [
        {
          givenname: "test",
          surname: "surname test",
          email: "test@email.es",
          phone: "12345",
          created_at: "",
        },
      ],
    });
  });

  test("renders correctly the component", async () => {
    const component = await factoryComponent();

    expect(ContactsService.show).toHaveBeenCalledWith(1);
    expect(component.asFragment()).toMatchSnapshot();
  });

  test("click on backBtn redirects correctly", async () => {
    const component = await factoryComponent();

    fireEvent.click(component.getByTestId("contact-show-back-btn"));

    expect(history.location.pathname).toBe("/contact-book/show/1");
  });

  test("click on editContact redirects correctly", async () => {
    const component = await factoryComponent();

    fireEvent.click(component.getByTestId("contact-show-edit-btn"));

    expect(history.location.pathname).toBe("/contact/1/edit");
  });
});
