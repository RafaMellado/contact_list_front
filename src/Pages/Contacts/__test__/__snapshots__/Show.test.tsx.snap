// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactShow /> renders correctly the component 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    <div
      class="mb-4 col-md-6 col-12"
    >
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <h2>
            test surname test
          </h2>
          <div
            class="mb-4"
          >
            <small
              class="text-secondary"
            >
              contactsShow.email
            </small>
            <h5>
              test@email.es
            </h5>
            <small
              class="text-secondary"
            >
              contactsShow.phone
            </small>
            <h5>
              12345
            </h5>
          </div>
          <button
            class="me-2 btn btn-primary"
            data-testid="contact-show-back-btn"
            type="button"
          >
            contactsShow.backBtn
          </button>
          <button
            class="me-2 btn btn-primary"
            data-testid="contact-show-edit-btn"
            type="button"
          >
            contactsShow.editContact
          </button>
        </div>
      </div>
    </div>
    <div
      class="col-md-6 col-12"
    >
      <h2>
        contactsShow.history
      </h2>
      <div
        class="mb-3 card"
        data-testid="contact-history-card"
      >
        <div
          class="card-body"
        >
          <h4>
            contactsShow.historyDay
          </h4>
          <small
            class="text-secondary"
          >
            contactsShow.name
          </small>
          <p>
            test surname test
          </p>
          <small
            class="text-secondary"
          >
            contactsShow.email
          </small>
          <p>
            test@email.es
          </p>
          <small
            class="text-secondary"
          >
            contactsShow.email
          </small>
          <p>
            12345
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
